File: backend/app/routes/auth_google_routes.py
from __future__ import annotations

from datetime import datetime, timezone
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import RedirectResponse
from sqlalchemy.orm import Session

from app.config.settings import settings
from app.database.session import get_db
from app.models.entities import ExternalAccount, User
from app.services.google_oauth import (
    build_google_auth_url_with_redirect,
    exchange_code_for_tokens_with_redirect,
    get_google_user_profile,
)
from app.utils.security import create_token, hash_password, decode_token


router = APIRouter(prefix="/auth", tags=["auth"])

GOOGLE_AUTH_SCOPES = [
    "openid",
    "email",
    "profile",
    # include these so Gmail/Calendar works immediately after Google sign-in
    "https://www.googleapis.com/auth/gmail.readonly",
    "https://www.googleapis.com/auth/gmail.modify",
    "https://www.googleapis.com/auth/calendar.readonly",
    "https://www.googleapis.com/auth/calendar.events",
    # classroom
    "https://www.googleapis.com/auth/classroom.courses.readonly",
    "https://www.googleapis.com/auth/classroom.coursework.me.readonly",
]


@router.get("/google/connect")
def google_auth_connect():
    if not settings.google_client_id or not settings.google_client_secret:
        raise HTTPException(status_code=500, detail="Google OAuth not configured on server")
    state = create_token({"type": "google_auth_state"}, 10)
    url = build_google_auth_url_with_redirect(
        state=state,
        scopes=GOOGLE_AUTH_SCOPES,
        redirect_uri=settings.google_auth_redirect_url,
    )
    return {"auth_url": url}


@router.get("/google/callback")
def google_auth_callback(
    code: str = Query(...),
    state: str = Query(...),
    db: Session = Depends(get_db),
):
    try:
        payload = decode_token(state)
        if payload.get("type") != "google_auth_state":
            raise ValueError("Invalid state")
    except Exception as exc:
        raise HTTPException(status_code=400, detail="Invalid OAuth state") from exc

    tokens = exchange_code_for_tokens_with_redirect(code=code, redirect_uri=settings.google_auth_redirect_url)
    profile = get_google_user_profile(access_token=tokens.access_token)
    email = str(profile.get("email") or "").strip().lower()
    name = str(profile.get("name") or "").strip() or (email.split("@")[0] if email else "User")

    if not email:
        raise HTTPException(status_code=400, detail="Google did not return an email")

    user = db.query(User).filter(User.email == email).first()
    if not user:
        # random password (user uses Google sign-in)
        user = User(name=name, email=email, hashed_password=hash_password(uuid4().hex))
        db.add(user)
        db.commit()
        db.refresh(user)

    # upsert external account so Gmail/Calendar is connected automatically
    ext = (
        db.query(ExternalAccount)
        .filter(ExternalAccount.user_id == user.id, ExternalAccount.provider == "google")
        .first()
    )
    if not ext:
        ext = ExternalAccount(user_id=user.id, provider="google")
        db.add(ext)
    ext.provider_user_email = email
    ext.access_token = tokens.access_token
    if tokens.refresh_token:
        ext.refresh_token = tokens.refresh_token
    ext.expires_at = tokens.expires_at
    ext.scope = tokens.scope
    ext.updated_at = datetime.now(timezone.utc)
    db.commit()

    access = create_token({"sub": str(user.id)}, settings.access_token_expire_minutes)
    refresh = create_token({"sub": str(user.id), "type": "refresh"}, settings.refresh_token_expire_minutes)

    # Send tokens back in URL hash so the browser doesn't send them to the server again.
    return RedirectResponse(
        url=f"{settings.frontend_origin}/oauth/google#access_token={access}&refresh_token={refresh}"
    )

File: backend/routes/authGoogleRoutes.js
import express from "express";
import crypto from "crypto";
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import User from "../models/User.js";
import ExternalAccount from "../models/ExternalAccount.js";
import { buildGoogleAuthUrl, exchangeCodeForTokens, fetchGoogleUserProfile } from "../services/googleService.js";

const router = express.Router();

const getAuthRedirectUri = () => {
  const backendOrigin = process.env.BACKEND_ORIGIN || "http://localhost:5000";
  return `${backendOrigin}/api/auth/google/callback`;
};

const createToken = (userId) => jwt.sign({ id: userId }, process.env.JWT_SECRET, { expiresIn: "7d" });

router.get("/google/start", async (req, res) => {
  if (!process.env.GOOGLE_CLIENT_ID || !process.env.GOOGLE_CLIENT_SECRET) {
    return res.status(500).json({ detail: "Google OAuth not configured on server" });
  }
  const frontendOrigin = process.env.FRONTEND_ORIGIN || "http://localhost:5173";
  const next = String(req.query?.next || "/dashboard");
  const state = Buffer.from(JSON.stringify({ next, t: Date.now(), nonce: crypto.randomUUID() }), "utf8").toString("base64url");
  const authUrl = buildGoogleAuthUrl({ state, redirectUri: getAuthRedirectUri() });
  return res.json({ auth_url: authUrl, frontend_origin: frontendOrigin });
});

router.get("/google/callback", async (req, res) => {
  const frontendOrigin = process.env.FRONTEND_ORIGIN || "http://localhost:5173";
  try {
    const code = String(req.query.code || "");
    const state = String(req.query.state || "");
    if (!code || !state) {
      return res.redirect(`${frontendOrigin}/login?google=error`);
    }

    const parsed = JSON.parse(Buffer.from(state, "base64url").toString("utf8"));
    const next = String(parsed?.next || "/dashboard");

    const tokens = await exchangeCodeForTokens(code, { redirectUri: getAuthRedirectUri() });
    const accessToken = String(tokens.access_token || "");
    if (!accessToken) {
      return res.redirect(`${frontendOrigin}/login?google=error`);
    }

    const profile = await fetchGoogleUserProfile(accessToken);
    const providerEmail = String(profile.email || "").trim().toLowerCase();
    const providerName = String(profile.name || providerEmail.split("@")[0] || "User").trim();

    if (!providerEmail) {
      return res.redirect(`${frontendOrigin}/login?google=error`);
    }

    let user = await User.findOne({ email: providerEmail });
    if (!user) {
      const randomPassword = crypto.randomBytes(32).toString("hex");
      const hashed = await bcrypt.hash(randomPassword, 10);
      user = await User.create({ name: providerName, email: providerEmail, password: hashed });
    }

    const account = await ExternalAccount.findOne({ userEmail: providerEmail, provider: "google" });
    const row = account || new ExternalAccount({ userEmail: providerEmail, provider: "google" });
    row.providerUserEmail = providerEmail;
    row.accessToken = accessToken;
    if (tokens.refresh_token) row.refreshToken = String(tokens.refresh_token);
    row.expiresAt = tokens.expires_in ? new Date(Date.now() + Number(tokens.expires_in) * 1000) : null;
    row.scope = String(tokens.scope || "");
    await row.save();

    const token = createToken(user._id);
    const redirectUrl = new URL(`${frontendOrigin}/oauth/google`);
    redirectUrl.searchParams.set("token", token);
    redirectUrl.searchParams.set("email", providerEmail);
    redirectUrl.searchParams.set("name", user.name || providerName);
    redirectUrl.searchParams.set("next", next);
    return res.redirect(redirectUrl.toString());
  } catch (_err) {
    return res.redirect(`${frontendOrigin}/login?google=error`);
  }
});

export default router;
